test/thirdparty.test.ts

```diff
@@ -0,0 +1,227 @@
+        },
+      ]);
+    });
+
+    test("basic query building works well to then query the data", async function () {
+      const { query, bindings } = GenerateQuery<User>(
+        QueryType.SELECT,
+        users.tableName, // Could also come from the above meta.txn objects
+        {
+          where: {
+            name: "Bobby Tables",
+          }, // this uses the type from above to enforce it to properties which exist on the table
+          limit: 1, // we only want the first user
+          offset: 1, // skip the first user named 'Bobby Tables' when performing this query
+
+          // Using orderBy is a special case, so there's a few possible syntaxes for it
+          orderBy: { column: "id", descending: true }, // ORDER BY id DESC NULLS LAST
+        }
+      );
+
+      // Using the database directly
+      const stmt = db.prepare(query).bind(bindings);
+      const { results } = await stmt.all();
+      deepStrictEqual(results, [
+        {
+          name: "Bobby Tables",
+          id: 1,
+          email: "bobby-tab@gmail.com",
+        },
+      ]);
+    });
+
+    test("where upserts are easier when using an orm", async function () {
+      const user: User = {
+        id: 1,
+        name: "John Doe",
+        email: "john-doe@gmail.com",
+      };
+      const { query, bindings } = GenerateQuery<User>(
+        QueryType.UPSERT,
+        users.tableName, // Could also come from the above meta.txn objects
+        {
+          data: user,
+          upsertOnlyUpdateData: {
+            name: user.name,
+            email: user.email,
+          },
+          where: {
+            id: user.id,
+          },
+        },
+        "id"
+      );
+      // {
+      //    query: "INSERT INTO users (id, name, email) VALUES (?, ?, ?) ON CONFLICT(id) DO UPDATE SET name = ?, email = ? WHERE id = ?",
+      //    bindings: [1, "John Doe", "john-doe@gmail.com", "John Doe", "john-doe@gmail.com", 1]
+      // }
+
+      // Using the database directly
+      const stmt = db.prepare(query).bind(bindings);
+      const { meta } = await stmt.run();
+      const receipt = await (meta.txn as any).wait();
+      strictEqual(receipt?.error, undefined);
+
+      const results = await db
+        .prepare(`SELECT * FROM ${users.tableName} WHERE id=?`)
+        .bind(user.id)
+        .first<User>();
+      deepStrictEqual(results, user);
+    });
+  });
+
+  describe("@databases/sql", function () {
+    // See https://www.atdatabases.org/docs/sqlite
+    const sqliteFormat: FormatConfig = {
+      escapeIdentifier: (str) => escapeSQLiteIdentifier(str),
+      formatValue: (value) => ({ placeholder: "?", value }),
+    };
+    let tableName: string;
+
+    this.beforeAll(async function () {
+      this.timeout("10s");
+
+      // First, we'll test out using sql identifiers
+      const primaryKey = sql.ident("id");
+      const query = sql`CREATE TABLE test_sql (${primaryKey} integer primary key, counter integer, info text);`;
```

Comment:
One nice thing about tagged template literals like this, is that you can actually get syntax highlighting. There is a plugin for VSCode, so we could also recommend that.